### ISSUE TITLE
Signout from Expired Session Forces you to login before logging out

### ISSUE THREAD

**Issue by parkinsona** [IdentityServer/IdentityServer3#1134]
Hi,

In the MVC katana sample, you use the following code to signout.
``` c#

        public ActionResult Signout()
        {
            Request.GetOwinContext().Authentication.SignOut();
            return Redirect("~/");
        }
```

If the user has a non-expired cookie for the RP, or Identity Server then all is good. It will log you out of both Idsvr & the Rp.
However, if you are expired in both the RP and Idsvr, it forces the user to login, and will then redirect them to the logout page.

I'm not sure if this is by design or not. I can see why it has to do this, but is there a way to just skip the logout?

I tried the following code. Which was handled this scenario.
However, if you were expired in the RP, but still active in Idsvr, it wouldn't log you out of idsvr.

``` c#
    [OverrideAuthorization]
        public ActionResult Signout()
        {
            if (Request.GetOwinContext().Authentication.User.Identity.IsAuthenticated)
            {
                Request.GetOwinContext().Authentication.SignOut();
            }

            //Request.GetOwinContext().Authentication.SignOut();
            return Redirect("~/");
        }

```

Is there a way around this behaviour?

**Comment by johnkors** [comment]
>However, if you are expired in both the RP and Idsvr, it forces the user to login, and will then redirect them to the logout page.

Is your logout endpoint at your RP requiring auth? The idsrv endession endpoint is not, AFAIK. Or do you do any redirects after idsrv-signout to a page requiring auth?

**Comment by parkinsona** [comment]
I have a global Authorize filter setup. So yes, it does require auth. Hence the [OverrideAuthorization] in my potential work around.

The page I redirect to does not require auth.

**Comment by johnkors** [comment]
ok, did you handle the ```OverrideAuthorization``` in  the global filter? A ```[AllowAnonymousAttribute]```for instance like  this:

 * Global filter
````
   public override void OnAuthorization(AuthorizationContext filterContext)
    {
        var skipAuthorization = filterContext.ActionDescriptor.IsDefined(typeof(AllowAnonymousAttribute), true) ||
                                filterContext.ActionDescriptor.ControllerDescriptor.IsDefined(
                                    typeof(AllowAnonymousAttribute), true);
        if (!skipAuthorization)
        {
            base.OnAuthorization(filterContext);
        }
    }
````

* ```AllowAnonymous``` attribute for your RP logout action/controller
```
[AttributeUsage(AttributeTargets.Method | AttributeTargets.Class, AllowMultiple = true, Inherited = true)]
public class AllowAnonymousAttribute : Attribute
{
}
```

Then :

```
        [AllowAnonymous]
        public ActionResult Signout()
        {
            
                Request.GetOwinContext().Authentication.SignOut();
         
        }
```

**Comment by parkinsona** [comment]
I did not handle the override in a filter.
I just used,  [OverrideAuthorization] as per my code sample above.

**Comment by brockallen** [comment]
Is this still an issue, or can we close?

**Comment by parkinsona** [comment]
@brockallen , this is still an issue. I've not found a solution yet.

**Comment by johnkors** [comment]
Did you try the approach of handling the override of authorization in your global filter, like i mentioned? It didn't work?

To me this seems not an idsrv issue, but a question of how auth mechanisms work in ASP.NET MVC/Katana (?)

**Comment by parkinsona** [comment]
Hi,

I never got a chance to try your suggestion, but I don't see how its different to doing the following:
``` c#
 [OverrideAuthorization]
        public ActionResult Signout()
        {
            Request.GetOwinContext().Authentication.SignOut();
            return Redirect("~/");
        }
```

From what I remember, when I tried this, it wouldn't sign me out of idsvr. I won't have time to try this out for a little while as I've been moved on to another project.

**Comment by johnkors** [comment]
It is different. Your global filter kicks in before your ```OverrideAuthorization``` attribute code runs. That's why I believe you have to add code to your global filter that excludes this. Look at the skipAuthorization bit in the global filter I pasted.

**Comment by leastprivilege** [comment]
[AllowAnonymous] is the right way to do it. A proper authZ attribute needs to check this.

**Comment by parkinsona** [comment]
Should I use [AllowAnonymous] for all unsecured pages or just the Signout action?

**Comment by parkinsona** [comment]
I've finally had a chance to look into this again.

I've got it working now. I actually got it working with both [AllowAnonymous] & [OverrideAuthorization].
I found that the global filter doesn't run if I user the [OverrideAuthoriation] attribute. [OverrideAuthentication] still runs the global filter.

Either way, I found where the issue was.
it's in the RedirectToIdentityProvider that adds the idTokentHint:

``` c#
 RedirectToIdentityProvider = async n =>
                    {
                        // if signing out, add the id_token_hint
                        if (n.ProtocolMessage.RequestType == OpenIdConnectRequestType.LogoutRequest)
                        {
                             //Needed to add a check to ensure the id_token exists. When the RP cookie had expired, the original code was throwing an exception.
                            var idToken = n.OwinContext.Authentication.User.FindFirst("id_token");
                            if (idToken != null)
                            {
                                var idTokenHint = idToken.Value;
                                n.ProtocolMessage.IdTokenHint = idTokenHint;
                            }
                        }
                    }
```

Might be worth updating the samples with this check.

**Comment by parkinsona** [comment]
I also found that using the [AllowAnonymous] attribute works, without any other modifications to the global filter's OnAuthorizationmethod. 

This makes sense with my current understanding of AllowAnonymous.

**Comment by johnkors** [comment]
I haven't tested the ```OverrideAuthentication``` nor ```OverrideAuthorization``` attributes. Seems they were introduced in ASP.NET MVC 5? I pasted a solution that we did on MVC3 back in the day, so I would not know - but I'm guessing the "new" MVC override stuff do more or less the same as the combination of ```AllowAnonymous```` and the skipping in the global filter. So the combo I pasted should work without using any Override-atts. At least for older MVC versions ;)

Glad to hear you got it working, though!
